master/buildbot/worker/upcloud.py

```diff
@@ -0,0 +1,167 @@
+        hostconfig = yield build.render(self.hostconfig)
+         # compose json
+        req = {
+            "server":{
+                "zone":hostconfig.get('zone', DEFAULT_ZONE),
+                "title":"Buildbot build #%d" % (self.substantiation_build.number,),
+                "hostname":hostconfig.get('hostname', self.name),
+                "user_data":hostconfig.get('user_data',""),
+                "login_user":{
+                    "username":"root",
+                    "ssh_keys": {
+                        "ssh_key": hostconfig.get('ssh_keys',[]),
+                    },
+                },
+                "storage_devices":{
+                    "storage_device":[{
+                        "action":"clone",
+                        "storage":image,
+                        "title":"Buildbot build #%d" % (self.substantiation_build.number,),
+                        "size":hostconfig.get("os_disk_size", DEFAULT_OS_DISK_SIZE),
+                        "tier":"maxiops",
+                    }],
+                }
+            }
+        }
+
+        req["server"]["plan"] = hostconfig.get("plan", DEFAULT_PLAN)
+        if req["server"]["plan"] == "custom":
+            req["server"]["core_number"] = hostconfig.get("core_number", DEFAULT_CORE_NUMBER)
+            req["server"]["memory_amount"] = hostconfig.get("memory_amount", DEFULT_MEMORY_AMOUNT)
+
+        # request instance
+        result = yield self.client.post("/server", json=req)
+
+        if int(result.code / 100) == 2:
+            instance = yield result.json()
+            self.instance = instance["server"]
+            self.instance["Id"] = self.instance["uuid"].split("-")[-1]
+            log.msg("UpcloudLatentWorker: Instance %s created (root password %s)" % (self.instance["Id"],self.instance['password']))
+            defer.returnValue([self.instance["Id"], image])
+        else:
+            self.failed_to_start(req['server']['hostname'], image)
+
+    @property
+    def shortid(self):
+        if self.instance is None:
+            return None
+        return self.instance['Id'][:6]
+
+    @defer.inlineCallbacks
+    def state(self):
+        result = yield self.client.get("/server/%s" % (self.instance["uuid"],))
+        if result.code == 404:
+            defer.returnValue("absent")
+        else:
+            server = yield result.json()
+            defer.returnValue(server["server"]["state"])
+
+    @defer.inlineCallbacks
+    def stop_instance(self, fast=False):
+        if self.instance is None:
+            # be gentle. Something may just be trying to alert us that an
+            # instance never attached, and it's because, somehow, we never
+            # started.
+            defer.succeed(None)
+            return
+        log.msg('UpcloudLatentWorker: Stopping instance %s...' % self.instance["Id"])
+        result = yield self.client.post("/server/%s/stop" % (self.instance["uuid"],), json={"stop_server":{"stop_type":"hard","timeout":"1"}})
```

Comment:
I would check that result, or we end up waiting forever and instance that couldn't stop.

Reply:
what should be done when the result is unexpected?

Reply:
at least log a failure case.

If the failure is because the instance does not exist, then you can exit soon.
else, probably a retry is worth a try.